### PR TITLE
Secret Service: cleanup and fix crash

### LINKED ISSUE #5279: Secret service crash when searching with null baseGroup
I experience the same on fresh install on Arch Linux. From `gdb`, I got:
```
ASSERT: "baseGroup" in file /home/invidian/repos/keepassxc/src/core/EntrySearcher.cpp, line 41
```

```
#0  0x00007ffff62a2355 in raise () from /usr/lib/libc.so.6
#1  0x00007ffff628b853 in abort () from /usr/lib/libc.so.6
#2  0x00007ffff6a429ac in QMessageLogger::fatal(char const*, ...) const () from /usr/lib/libQt5Core.so.5
#3  0x00007ffff6a41d59 in qt_assert(char const*, char const*, int) () from /usr/lib/libQt5Core.so.5
#4  0x0000555555792888 in EntrySearcher::search (this=0x7fffffffd5a0, searchTerms=..., baseGroup=0x0, forceSearch=false) at /home/invidian/repos/keepassxc/src/core/EntrySearcher.cpp:41
#5  0x00005555557fd8a9 in FdoSecrets::Collection::searchItems (this=0x555558571320, attributes=...) at /home/invidian/repos/keepassxc/src/fdosecrets/objects/Collection.cpp:254
#6  0x00005555557f20ca in FdoSecrets::Service::searchItems (this=0x55555667f3f0, attributes=..., locked=...) at /home/invidian/repos/keepassxc/src/fdosecrets/objects/Service.cpp:267
#7  0x000055555580ae7c in FdoSecrets::ServiceAdaptor::SearchItems (this=0x55555667f730, attributes=..., locked=...) at /home/invidian/repos/keepassxc/src/fdosecrets/objects/adaptors/ServiceAdaptor.cpp:71
#8  0x00005555557de07a in FdoSecrets::ServiceAdaptor::qt_static_metacall (_o=0x55555667f730, _c=QMetaObject::InvokeMetaMethod, _id=5, _a=0x7fffffffd8c0)
    at /home/invidian/repos/keepassxc/build/src/fdosecrets/fdosecrets_autogen/ZPMS5EPCTP/moc_ServiceAdaptor.cpp:155
#9  0x00005555557de76b in FdoSecrets::ServiceAdaptor::qt_metacall (this=0x55555667f730, _c=QMetaObject::InvokeMetaMethod, _id=5, _a=0x7fffffffd8c0)
    at /home/invidian/repos/keepassxc/build/src/fdosecrets/fdosecrets_autogen/ZPMS5EPCTP/moc_ServiceAdaptor.cpp:313
#10 0x00007ffff6f25fd6 in ?? () from /usr/lib/libQt5DBus.so.5
#11 0x00007ffff6f295cb in ?? () from /usr/lib/libQt5DBus.so.5
#12 0x00007ffff6f29eba in ?? () from /usr/lib/libQt5DBus.so.5
#13 0x00007ffff6f2c388 in ?? () from /usr/lib/libQt5DBus.so.5
#14 0x00007ffff6c911d2 in QObject::event(QEvent*) () from /usr/lib/libQt5Core.so.5
#15 0x00007ffff77ee702 in QApplicationPrivate::notify_helper(QObject*, QEvent*) () from /usr/lib/libQt5Widgets.so.5
#16 0x00007ffff6c647ba in QCoreApplication::notifyInternal2(QObject*, QEvent*) () from /usr/lib/libQt5Core.so.5
#17 0x00007ffff6c672a3 in QCoreApplicationPrivate::sendPostedEvents(QObject*, int, QThreadData*) () from /usr/lib/libQt5Core.so.5
#18 0x00007ffff6cbdcf4 in ?? () from /usr/lib/libQt5Core.so.5
#19 0x00007ffff544a43c in g_main_context_dispatch () from /usr/lib/libglib-2.0.so.0
#20 0x00007ffff54981d9 in ?? () from /usr/lib/libglib-2.0.so.0
#21 0x00007ffff5449221 in g_main_context_iteration () from /usr/lib/libglib-2.0.so.0
#22 0x00007ffff6cbd331 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/libQt5Core.so.5
#23 0x00007ffff6c6313c in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/libQt5Core.so.5
#24 0x00007ffff6c6b5c4 in QCoreApplication::exec() () from /usr/lib/libQt5Core.so.5
#25 0x00005555555c5919 in main (argc=1, argv=0x7fffffffe258) at /home/invidian/repos/keepassxc/src/main.cpp:160
```

When trying to execute the following command:
```sh
secret-tool lookup UserName invidian
```

Version: 71b05dbcf48efbeb2341f28eac697aaa1616aeb2.

_Originally posted by @invidian in https://github.com/keepassxreboot/keepassxc/issues/5199#issuecomment-674551220_

### PR DESCRIPTION
**Pull request by Aetf**
[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Explain large or complex code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" )
Fixes #5279. Aside from typos and formatting fixing, this does the following

- Correctly propagate error message to the UI rather than ignoring them
- Simplify the internal states of `Collection`, so it's less likely to be left in an inconsistent state
- Handle corner cases when registering Collection on DBus
  * Previously `QFileInfo::baseName` was used to derive the name, but this can be empty if there are leading dots in the filename, causing #5279. `QFileInfo::completeBaseName` is used instead.
  * Although unlikely, if two open databases have the same filename (but on different paths), the second database will fail to register. New logic is added to address this.

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )
New test cases added covering the corner cases mentioned above.

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ Bug fix (non-breaking change that fixes an issue)
- ✅ Refactor (significant modification to existing code)

### CODE REVIEW

File: tests/gui/TestGuiFdoSecrets.cpp [hunk 1]
```diff
@@ -761,6 +762,47 @@ void TestGuiFdoSecrets::testCollectionDelete()
     }
 }
 
+void TestGuiFdoSecrets::testHiddenFilename()
+{
+    // when file name contains leading dot, all parts excepting the last should be used
+    // for collection name, and the registration should success
+    QVERIFY(m_dbFile->rename(QFileInfo(*m_dbFile).path() + "/.Name.kdbx"));
+
+    m_db = nullptr;
```
Comment (droidmonkey) [inline tests/gui/TestGuiFdoSecrets.cpp:771]: Must be `m_db.reset()`, technically unnecessary since you assign the pointer below anyway.

**Review by droidmonkey — CHANGES_REQUESTED** [review]

File: src/fdosecrets/objects/Collection.cpp [hunk 2]
```diff
@@ -95,6 +98,20 @@ namespace FdoSecrets
         } else {
             cleanupConnections();
         }
+
+        return true;
+    }
+
+    QDBusAbstractAdaptor* Collection::createAdaptor()
+    {
+        return new CollectionAdaptor(this);
```
Comment (droidmonkey) [inline src/fdosecrets/objects/Collection.cpp:107]: This is a great way to leak memory.

**Review by droidmonkey — CHANGES_REQUESTED** [review]

File: tests/gui/TestGuiFdoSecrets.cpp (hunk 1 above)
Comment (Aetf) [reply tests/gui/TestGuiFdoSecrets.cpp:771]: Yeah noticed the build failure. It necessary though, otherwise the old database is not deleted and connections not removed until it's assigned below, at which point other things have changed and it's causing some problem.

File: src/fdosecrets/FdoSecretsPlugin.cpp [hunk 3]
```diff
@@ -65,12 +63,8 @@ void FdoSecretsPlugin::updateServiceState()
...
     if (FdoSecrets::settings()->isEnabled()) {
         if (!m_secretService && m_dbTabs) {
-            m_secretService.reset(new Service(this, m_dbTabs));
-            connect(m_secretService.data(), &Service::error, this, [this](const QString& msg) {
-                emit error(tr("<b>Fdo Secret Service:</b> %1").arg(msg));
-            });
-            if (!m_secretService->initialize()) {
-                m_secretService.reset();
+            m_secretService = Service::Create(this, m_dbTabs);
+            if (!m_secretService) {
```
[the remark this answers is not in the document]
Comment (Aetf) [reply src/fdosecrets/FdoSecretsPlugin.cpp:67]: It's a little confusing. But my principle in at least `fdosecrets/*` is to report error messages at the exact place where the error happens (See all the usage sites of `FdoSecretsPlugin::emitError`). So methods returning `nullptr` or `false` should have printed the message already.

File: src/fdosecrets/objects/Collection.cpp (hunk 2 above)
Comment (Aetf) [reply src/fdosecrets/objects/Collection.cpp:107]: Adaptors are parented to `this` and will only be created once.

File: src/fdosecrets/objects/Collection.cpp [hunk 4]
```diff
@@ -34,16 +34,20 @@
 
 namespace FdoSecrets
 {
+    Collection* Collection::Create(Service* parent, DatabaseWidget* backend)
+    {
+        std::unique_ptr<Collection> coll{new Collection(parent, backend)};
```
[the remark this answers is not in the document]
Comment (Aetf) [reply src/fdosecrets/objects/Collection.cpp:39]: They are mostly the same until not :/ I noticed `QScopedPointer` doesn't have move-semantics and can't be simply returned in `Service::Create`. Others are just for uniformity.

File: src/fdosecrets/objects/Collection.cpp (hunk 2 above)
Comment (Aetf) [reply src/fdosecrets/objects/Collection.cpp:107]: I'm open to how to make the intent more clear

Comment (Aetf) [reply src/fdosecrets/objects/Collection.cpp]: Now the creation of the adaptor is enforced in the constructor.

File: src/fdosecrets/objects/Collection.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (phoerious) [reply src/fdosecrets/objects/Collection.cpp:39]: If you need to return a pointer, use QSharedPointer.

File: src/fdosecrets/objects/Collection.cpp (hunk 2 above)
Comment (droidmonkey) [reply src/fdosecrets/objects/Collection.cpp]: Use a QSharedPointer

File: src/fdosecrets/objects/Collection.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Aetf) [reply src/fdosecrets/objects/Collection.cpp:39]: Isn't using a QSharedPointer dangerous for objects with a parent? I.e. if someone holds the shared pointer longer than the lifetime of the parent object, the shared pointer will try to delete the object again.

IMHO QSharedPointer represents a completely different ownership mechanism than QObject's parent/child mechanism and should not be used at the same time.

All these dbus objects are managed by QObject's parent/child mechanism except Service, which can't be parented to FdoSecretPlugin due to some complications in DBusObject's interface.

The `*::Create` methods are introduced to allow failable object constructions. And the std::unique_ptr/QScopedPointer in them is to ensure in the failure case, the object is properly deleted before returning. They still produce objects managed by QObject's parent/child mechanism (except Service) and that's why they return raw pointers.

File: src/fdosecrets/objects/Collection.cpp (hunk 2 above)
Comment (droidmonkey) [reply src/fdosecrets/objects/Collection.cpp]: We only use the parent child relationship for QWidgets. It causes problems unless you use QPointer to reference the children because you scatter dangling pointers around when the parent deletes.

Comment (Aetf) [reply src/fdosecrets/objects/Collection.cpp]: I understand. In this particular case, the adaptor has to be the child otherwise Qt DBus won't find it. Anyway, this method is gone now and I create these adaptors directly in the DBusObjectHelper constructor.

As for others (Service, Collection, Item, Session), it's indeed possible and maybe more appropriate to move away from the parent child relationship. But let's save it for another PR.

Comment (droidmonkey) [reply src/fdosecrets/objects/Collection.cpp]: Sounds good to me

**Comment by Aetf** [comment]
The CI failure on Windows seems unrelated

**Comment by droidmonkey** [comment]
@aetf Is this ready to go?

**Comment by Aetf** [comment]
@droidmonkey Yes,  it's ready. I just rebased.

**Comment by Aetf** [comment]
@droidmonkey ping. I also fixed a signal connection problem that was introduced during the refactoring.

**Review by droidmonkey — APPROVED** [review]

**Comment by droidmonkey** [comment]
Excellent! Can we merge the prompt for access one soon?

**Comment by Aetf** [comment]
I'm working on that one now. It still needs some work and doesn't apply cleanly with this merged. I'll update very soon 😄
